lib/base/Version.ts

```diff
@@ -250,42 +271,69 @@ export default class Version {
     let currentPage = 1;
     let currentResource = 0;
     let limits = {} as PageLimit;
+    let pPending = true;
+    let pResolve: (value: void) => void;
+    let pReject: (reason?: any) => void;
     if (this._version instanceof Version) {
       limits = this._version.readLimits({
         limit: params.limit,
         pageSize: params.pageSize,
       });
     }
-    function onComplete(error?) {
+    function onComplete(error?: any) {
+      let unhandledError = error;
+
       done = true;
       if (typeof params.done === "function" && !doneCalled) {
-        params.done(error);
+        try {
+          params.done(unhandledError);
+          unhandledError = null;
+        } catch (e) {
+          unhandledError = e;
+        }
       }
       doneCalled = true;
+
+      if (pPending) {
+        if (unhandledError) {
+          pReject(unhandledError);
+        } else {
+          pResolve();
+        }
+        pPending = false;
+      }
     }
-    function fetchNextPage(fn) {
+    function fetchNextPage(fn: () => Promise<any>) {
       let promise = fn();
       if (typeof promise === "undefined") {
         onComplete();
         return;
       }
+
       promise.then((page) => {
-        page.instances.forEach(function (instance: any) {
-          if (
-            done ||
-            (typeof params.limit !== "undefined" &&
-              currentResource >= params.limit)
-          ) {
-            done = true;
-            return false;
-          }
-          currentResource++;
-          try {
-            callback(instance, onComplete);
-          } catch (e) {
-            throw e;
-          }
-        });
+        try {
+          page.instances.forEach(function (instance: any) {
+            if (
+              done ||
+              (typeof params.limit !== "undefined" &&
+                currentResource >= params.limit)
+            ) {
+              done = true;
+              return false;
+            }
+            currentResource++;
+            try {
+              if (callback) {
```

Comment:
Shouldn't a callback always be defined?

Reply:
Yes, I was narrowing it to avoid a tsc error. It looks like the compiler isn't able to understand that `callback` can never be undefined thanks to the conditional check and exception thrown above. I changed this to use an [optional chaining operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html), which seems to also avoid the compiler error.